snowpark_style_guide.md

```diff
@@ -0,0 +1,341 @@
+roles the parameters play instead of constantly referencing to the API doc to figure out.
+
+```python
+# without parameter name
+df = df.join(another_df, 'id', 'inner')
+df = df.unpivot("sales", "month", ["jan", "feb"]).sort("empid")
+
+# with parameter name
+df = df.join(another_df, using_columns='id', how='inner')
+df = df.unpivot(value_column="sales", name_column="month", column_list=["jan", "feb"]).sort("empid")
+```
+
+# User Defined Functions (UDFs)
+
+[User Defined Functions (UDFs)][udf-service-doc] let you extend the system to perform customized operations and Snowpark
+lets you use the Python programming language to manipulate data and return either scalar or tabular results.
+
+UDFs could be registered in several ways and each has its own advantages and limitations.
+
+
+## Registration
+
+**calling `snowflake.snowpark.functions.udf` function**
+
+The `snowflake.snowpark.functions.udf` function will register the function within the most
+recently created session. If there is only one single active session in the program, it will use that session
+for registration. However, if there are multiple active sessions in your
+program and a different session is wanted, the `session` parameter has to be explicitly set.
+In this case, you may need to keep a record of all the sessions and maintain extra logic to decide on which
+session to use.
+
+```python
+from snowflake.snowpark.functions import udf
+
+def add(x, y):
+     return x + y
+
+# without session parameter, the most recently created session will be used
+udf(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+
+# or explicitly specifying an active session
+udf(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()],
+   session=active_session
+)
+```
+
+**using `snowflake.snowpark.functions.udf` as decorator**
+
+`snowflake.snowpark.functions.udf` could be also used as a decorator to annotate the function to be registered.
+With the decorator, it is clearer whether the function is a UDF.
+
+```python
+# udf as decorator
+from snowflake.snowpark.functions import udf
+
+@udf(
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+def add(x, y):
+     return x + y
+```
+
+**calling `snowflake.snowpark.Session.udf.register` method**
+
+The `udf` method on `Session` object returns `UDFRegistration` object which could be used for registration.
+Inherently, `UDFRegistration` is bound with the `Session` object which means the UDF will be registered
+through that session.
+
+```python
+def add(x, y):
+     return x + y
+
+session.udf.register(
+   add,
+   return_type=IntegerType(),
+   input_types=[IntegerType(), IntegerType()]
+)
+```
+
+## Type Hints
+
+As illustrated in the above [UDF Registration][#UDF Registration], the typing info needs to be provided.
+Other than inputting types as parameters, you could also provide the typing information through Python type hints.
```

Comment:
Let's mention several snowpark-defined type annotation in snowflake.snowpark.types (Variant, Geography, PandasSeries, PandasDataFrame)

Reply:
updated with sample